8247281: migrate ObjectMonitor::_object to OopStorage

dcubed-ojdk opened the pull request:
This RFE is to migrate the following field to OopStorage:

class ObjectMonitor {
<snip>
  void* volatile _object; // backward object pointer - strong root

Unlike the previous patches in this series, there are a lot of collateral
changes so this is not a trivial review. Sorry for the tedious parts of
the review. Since Erik and I are both contributors to this patch, we
would like at least 1 GC team reviewer and 1 Runtime team reviewer.

This changeset was tested with Mach5 Tier[1-3],4,5,6,7,8 testing
along with JDK-8252980 and JDK-8252981. I also ran it through my
inflation stress kit for 48 hours on my Linux-X64 machine.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8247281](https://bugs.openjdk.java.net/browse/JDK-8247281): migrate ObjectMonitor::_object to OopStorage


### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to eeb9d761374ea193a7295fcce3ffe9707a8f0348
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 750fe771943178a02f1b71a713f7417512e3628e
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to eeb9d761374ea193a7295fcce3ffe9707a8f0348
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)


### Contributors
 * Erik Österlund `<erik.osterlund@oracle.com>`
 * Daniel Daugherty `<daniel.daugherty@oracle.com>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/135/head:pull/135`
`$ git checkout pull/135`

dcubed-ojdk commented:
/contributor add Erik Österlund erik.osterlund@oracle.com
/contributor add Daniel Daugherty daniel.daugherty@oracle.com

dcubed-ojdk commented:
Erik Österlund <erik.osterlund@oracle.com>
Daniel Daugherty <daniel.daugherty@oracle.com>

dcubed-ojdk commented:
/label remove serviceability,hotspot
/label add hotspot-runtime,hotspot-gc

dcubed-ojdk commented:
@fisk - Please chime in on this review when you get the chance.

dcubed-ojdk reviewed (commented):
The vast majority of these changes came from @fisk so I'm actually a
reviewer and stress tester of these changes. Thumbs up!

fisk reviewed (approved):
This looks great Dan. I like your addition of is_chainmarker() - it makes this step a lot better, without having any special oops that are not oops. Thanks for sorting this out!

dcubed-ojdk commented:
@fisk - Thanks for the blinding fast review! (Pretty easy when you wrote almost
all of the code).

Re: is_chainmarker()
I figured you would appreciate getting rid of one more "special" oop value!
And its use just fits in with the whole AllocationState model. It also gets
removed quite easily with the part3 patch... Oh yeah, I gotta file a new RFE
for that one.

coleenp commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -159,8 +159,6 @@ struct ObjectMonitorListGlobals {
 static ObjectMonitorListGlobals om_list_globals;
 
-#define CHAINMARKER (cast_to_oop<intptr_t>(-1))
-
```
Great. One less non-oop oop!

dcubed-ojdk replied:
Exactly!

fisk commented on src/hotspot/share/prims/jvmtiTagMap.cpp:
```diff
@@ -3017,13 +3017,6 @@ inline bool VM_HeapWalkOperation::collect_simple_roots() {
   }
 
-  // Inflated monitors
-  blk.set_kind(JVMTI_HEAP_REFERENCE_MONITOR);
```
The JVMTI roots are strong roots. This is no longer a strong root, so reporting them would be a bug after this change.

dcubed-ojdk replied:
The function is VM_HeapWalkOperation::collect_simple_roots()
and we no longer have a root in the ObjectMonitor so my take is
no we don't. I believe @fisk concurs with that reasoning.

fisk replied:
Yes, exactly.

dcubed-ojdk replied:
Thanks for confirmation.

dholmes-ora replied:
From the spec I'm not clear on exactly what JVMTI_HEAP_REFERENCE_MONITOR is intended to be. Serviceability folk should be giving some input here though.

dcubed-ojdk replied:
I've taken a first pass at creating a CSR:
JDK-8253121 migrate ObjectMonitor::_object to OopStorage
https://bugs.openjdk.java.net/browse/JDK-8253121

sspitsyn replied:
Just a minor comment.
The line 1754 can be deleted as the JVMTI_HEAP_REFERENCE_MONITOR reference type will be never encountered:

1750 static jvmtiHeapRootKind toJvmtiHeapRootKind(jvmtiHeapReferenceKind kind) {
1751   switch (kind) {
1752     case JVMTI_HEAP_REFERENCE_JNI_GLOBAL:   return JVMTI_HEAP_ROOT_JNI_GLOBAL;
1753     case JVMTI_HEAP_REFERENCE_SYSTEM_CLASS: return JVMTI_HEAP_ROOT_SYSTEM_CLASS;
1754     case JVMTI_HEAP_REFERENCE_MONITOR:      return JVMTI_HEAP_ROOT_MONITOR;
1755     case JVMTI_HEAP_REFERENCE_STACK_LOCAL:  return JVMTI_HEAP_ROOT_STACK_LOCAL;
1756     case JVMTI_HEAP_REFERENCE_JNI_LOCAL:    return JVMTI_HEAP_ROOT_JNI_LOCAL;
1757     case JVMTI_HEAP_REFERENCE_THREAD:       return JVMTI_HEAP_ROOT_THREAD;
1758     case JVMTI_HEAP_REFERENCE_OTHER:        return JVMTI_HEAP_ROOT_OTHER;
1759     default: ShouldNotReachHere();          return JVMTI_HEAP_ROOT_OTHER;
1760   }
1761 }

Other than that the update in this file looks okay to me.

dcubed-ojdk replied:
I cleaned that up. The only references to JVMTI_HEAP_REFERENCE_MONITOR and
JVMTI_HEAP_ROOT_MONITOR that remain are in the spec:

$ egrep -r 'JVMTI_HEAP_REFERENCE_MONITOR|JVMTI_HEAP_ROOT_MONITOR' src/hotspot
src/hotspot/share/prims/jvmti.xml:          <constant id="JVMTI_HEAP_REFERENCE_MONITOR" num="23">
src/hotspot/share/prims/jvmti.xml:          <constant id="JVMTI_HEAP_ROOT_MONITOR" num="3">
src/hotspot/share/prims/jvmti.xml:              <code>JVMTI_HEAP_ROOT_MONITOR</code>,

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -237,6 +241,49 @@ void ObjectMonitor::operator delete[] (void *p) {
 
+// Check that object() and set_object() are called from the right context:
+static void check_object_context() {
+  Thread *self = Thread::current();
```
I'll fix that.

dcubed-ojdk replied:
Fixed in https://github.com/openjdk/jdk/pull/135/commits/750fe771943178a02f1b71a713f7417512e3628e

dcubed-ojdk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -237,6 +241,49 @@ void ObjectMonitor::operator delete[] (void *p) {
+  // the VMThread so sanity check it.
+  guarantee(self->is_Java_thread() || self->is_VM_thread(), "must be");
+  if (self->is_Java_thread()) {
+    JavaThread* jt = (JavaThread*)self;
```
Yup. Since this is a new function, it didn't pop up as a conflict.
I'll fix that.

dcubed-ojdk replied:
Fixed in https://github.com/openjdk/jdk/pull/135/commits/750fe771943178a02f1b71a713f7417512e3628e.

fisk commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -237,6 +241,49 @@ void ObjectMonitor::operator delete[] (void *p) {
+  if (_object.is_null()) {
+    return NULL;
+  }
+  return _object.resolve();
```
Because before the TSM removal patch, the monitors are allocated in blocks. Then, the objects are not known when the monitors are allocated. The handles are instead lazily created... for now. With the next patch, that will probably change, if we agree to allocate ObjectMonitors one by one on inflation time. But that is one patch away.

dcubed-ojdk replied:
Erik's part3 patch ensures that when the ObjectMonitor is allocated,
the weak handle is created and initialized to the oop at that point. Since
this is still the TSM version, the ObjectMonitor may not yet have a weak
handle allocated at the time that object() is called.

coleenp replied:
Ok, this is cool then and good that this logic is here instead of inside the WeakHandle::resolve/peek functions.  The OopHandle similar functions have a null check, so if this was permanent, I'd suggest putting it there.  Since this is going to change, having it here and keeping the non-null invariants for WeakHandle::resolve/peek functions is better.

dcubed-ojdk replied:
Thanks for confirmation.

fisk commented on src/hotspot/share/services/heapDumper.cpp:
```diff
@@ -1379,27 +1379,6 @@ void JNIGlobalsDumper::do_oop(oop* obj_p) {
-  }
-  void do_oop(oop* obj_p) {
-    u4 size = 1 + sizeof(address);
-    writer()->start_sub_record(HPROF_GC_ROOT_MONITOR_USED, size);
```
Same answer as above. No longer a strong root, so can't report it as such.

dcubed-ojdk replied:
We no longer have a root in the ObjectMonitor so no we don't have
to dump these is my take. I believe @fisk concurs with that reasoning.

fisk replied:
Agreed.

coleenp replied:
Ok, that's good then!

dcubed-ojdk replied:
Thanks for confirmation.

dcubed-ojdk replied:
I've taken a first pass at creating a CSR:
JDK-8253121 migrate ObjectMonitor::_object to OopStorage
https://bugs.openjdk.java.net/browse/JDK-8253121

sspitsyn replied:
I've looked at the CSR and added myself as a reviewer.
We had a slack chat with Dan, and I agree that with a weak handle it would be racy/unsafe for JVMTI_HEAP_REFERENCE_MONITOR calls back to be called. The ObjectMonitors do not pin objects anymore (there are no strong refs from them), so it has to be okay to skip reporting the JVMTI_HEAP_REFERENCE_MONITOR and and JVMTI_HEAP_ROOT_MONITOR (old Heap Walking API) reference types. The JVMTI does not need an update as other VM implementations can still report these reference types. Alan added a comment to the CSR saying that memory profiling tools that use the JVMTI functions (FollowReferences with jvmtiHeapReferenceCallback or IterateOverReachableObjects with jvmtiHeapRootCallback) to iterate over the heap should not have a compatibility impact as these reference types are just informational but adding a release note can be useful.

dcubed-ojdk replied:
Slight clarification. Serguei and I were discussing whether we could continue
to make JVMTI_HEAP_REFERENCE_MONITOR call backs or emit
HPROF_GC_ROOT_MONITOR_USED entries in heap dump output as a
way to ease the transition phase of getting used to these things going
away. My answer was that we could do that but it would racy and unsafe
due to the ObjectMonitor's object being GC'ed.

It's also incorrect to make JVMTI_HEAP_REFERENCE_MONITOR call backs
or emit HPROF_GC_ROOT_MONITOR_USED entries in heap dump once
the ObjectMonitor's object ref becomes a weak handle. That weak handle
no longer prevents the associated object from being GC'ed so it is no
longer a strong root.

See Erik's comment above: https://github.com/openjdk/jdk/pull/135#discussion_r487291746

dcubed-ojdk commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -1575,7 +1545,7 @@ ObjectMonitor* ObjectSynchronizer::om_alloc(Thread* self) {
                                     bool from_per_thread_alloc) {
   guarantee(m->header().value() == 0, "invariant");
-  guarantee(m->object() == NULL, "invariant");
+  guarantee(m->object_peek() == NULL, "invariant");
```
Correct. See:
+void ObjectMonitor::set_object(oop obj) {
+  check_object_context();
+  if (_object.is_null()) {
+    _object = WeakHandle(_oop_storage, obj);
+  } else {
+    _object.replace(obj);
+  }
+}

So when 'obj' == NULL, we replace the weak handle's value with NULL
and we don't release/delete/whatever the weak handle.

dcubed-ojdk replied:
I can't figure out what "magic" to use to get the above code quote to show up as such.

coleenp commented on src/hotspot/share/runtime/thread.cpp:
```diff
@@ -4595,6 +4584,13 @@ void Threads::remove(JavaThread* p, bool is_daemon) {
+  // This must be done after ObjectSynchronizer::om_flush(), as GC barriers
+  // are used in om_flush().
+  BarrierSet::barrier_set()->on_thread_detach(p);
+
```
One last question, that doesn't require a comment why it's here, but why was this moved?

dcubed-ojdk replied:
We had to move that code to make om_flush() happy.
om_flush() accesses the (weak) oops in the monitor list
after the thread is off the threads list so the barrier change
had to move. It will move back in part3.

dcubed-ojdk replied:
Sorry, I should have been more clear here.
We had to move that code after the om_flush() call because om_flush()
accesses the (weak) oops in the monitor list after the original location
of the code. The om_flush() call is made in the code that removes the
thread from the threads list, but the removal happens after the call to
om_flush().

coleenp reviewed (approved):
I approve of this without seeing any new changes.

coleenp commented:
I have to also comment that reviewing code in github is really nice!

dcubed-ojdk commented:
Agreed. When I did my self-review for this fix, I didn't use the webrev link at all.
I did generate a local webrev via "git webrev" and sanity checked that webrev
to verify that "git webrev" works (and it did).

dholmes-ora reviewed (commented):
Overall cleanup looks good. More complicated than I had expected, but then this changes the GC races with the deflation code.

Not sure about the JVM TI changes!

dholmes-ora commented:
Please note that I made comments on the "resolved" discussion on heapDumper.cpp  and jvmtiTagMap.cpp  but they were not mentioned in the review email and are hidden in the UI unless you explicitly click on "Show resolved".

stefank reviewed (approved):
GC changes looks good to me.

rkennke commented:
Shenandoah changes are not complete:
/home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp: In member function 'virtual void ShenandoahFinalMarkingTask::work(uint)':
/home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:298:31: error: 'oops_do' is not a member of 'ObjectSynchronizer'
           ObjectSynchronizer::oops_do(&resolve_mark_cl);
                               ^~~~~~~
/home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:308:31: error: 'oops_do' is not a member of 'ObjectSynchronizer'
           ObjectSynchronizer::oops_do(&mark_cl);
                               ^~~~~~~

I will have a look into how to resolve this.

dcubed-ojdk commented:
@stefank - Thanks for the review!

dcubed-ojdk commented:
@dholmes-ora - Thanks for the review. I've think I have made all the comments
you were concerned about as unresolved. I agree that we need a CSR for the
change in behavior and I'll look into getting that going.

dcubed-ojdk commented:
@kimbarrett - Thanks for the review. I believe @fisk is going to
address your comments.

dcubed-ojdk commented:
@rkennke - Thanks for the review. I believe @fisk is going to address
your comments.

dcubed-ojdk commented:
/label add serviceability

fisk commented:
Hi Kim,

Here is a partial reply to your review. Thanks for reviewing!
Hmm seems like your email was only sent to shenandoah-dev. Not sure if that was intended. I'm not subscribed to that mailing list, so I will send my reply through github and hope for the best.

> _Mailing list message from [Kim Barrett](mailto:kim.barrett@oracle.com) on [shenandoah-dev](mailto:shenandoah-dev@openjdk.java.net):_
> 
> ------------------------------------------------------------------------------
> src/hotspot/share/oops/weakHandle.cpp
> 36 WeakHandle::WeakHandle(OopStorage* storage, oop obj) :
> 37 _obj(storage->allocate()) {
> 38 assert(obj != NULL, "no need to create weak null oop");
> 
> Please format this differently so the ctor-init-list is more easily
> distinguished from the body. I don't care that much which of the several
> alternatives is used.
> 
> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/objectMonitor.cpp
> 244 // Check that object() and set_object() are called from the right context:
> 245 static void check_object_context() {
> 
> This seems like checking we would normally only do in a debug build. Is this
> really supposed to be done in product builds too? (It's written to support
> that, just wondering if that's really what we want.) Maybe these aren't
> called very often so it doesn't matter? I also see that guarantee (rather
> than assert) is used a fair amount in this and related code.

I don't think I have a preference here. As you say, it seems a bit mixed. I would be okay with both. Do you want them to be asserts?

> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/objectMonitor.cpp
> 251 JavaThread* jt = (JavaThread*)self;
> 
> Use self->as_Java_thread().
> 
> Later: Coleen already commented on this and it's been fixed.
> 
> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/objectMonitor.cpp
> 249 guarantee(self->is_Java_thread() || self->is_VM_thread(), "must be");
> 250 if (self->is_Java_thread()) {
> 
> Maybe instead
> 
> if (self->is_Java_thread()) {
> ...
> } else {
> guarantee(self->is_VM_thread(), "must be");
> }
> 
> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/objectMonitor.cpp
> Both ObjectMonitor::object() and ObjectMonitor::object_peek() have
> 265 if (_object.is_null()) {
> 266 return NULL;
> 
> Should we really be calling those functions when in such a state? That seems
> like it might be a bug in the callers?
> 
> OK, I think I see some places where object_peek() might need such protection
> because of races, in src/hotspot/share/runtime/synchronizer.cpp. And
> because we don't seem to ever release() the underlying WeakHandles.
> 
> 514 if (m->object_peek() == NULL) {
> 703 if (cur_om->object_peek() == NULL) {
> 
> But it still seems like it might be a bug to call object() in such a state.

The problem is Type Stable Memory (TSM). The idea is that all monitors are allocated in blocks of many (when the object they are associated with is not yet known), and are then re-used, switching object association back and forth. They are also never freed.

The changes in this RFR were originally tied together with a patch that also removes TSM. But this has been split into separate parts by Dan in order to perform the OopStorage migration first, and then removing TSM. So in this patch, we are still dealing with block-allocation of monitors before they have an associated object, and changing of object association. Therefore, what you are saying is 100% true, but I would like to defer fixing that until the follow-up patch that removes TSM. In that patch, the accessors are changed to do exactly what you expect them to (resolve/peek), as the ObjectMonitor instances are allocated with a single object association in mind, installed in the constructor, and get deleted after deflation, with destuctors run. So I hope we can agree to wait until that subsequent patch with addressing such concerns. We thought reviewing the whole thing as a single patch was too much clutter.

> Related, see next comment.
> 
> Later: Looks like Coleen questioned this too. I'm not sure I understand
> Erik's response though. When do we look at monitors that might not be
> constructed / initialized? That seems like a bad thing to do. Oh, but the
> whole creation path for OMs is kind of evil right now. I see... And that's
> planned to be fixed in later work. OK.

Yes, exactly. It is currently quite evil, but will soon be cleaned up to look quite ordinary.

> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/synchronizer.cpp
> 1548 guarantee(m->object_peek() == NULL, "invariant");
> 
> Later: But see previous comment. Some of this might be relevat later though.
> 
> object_peek() seemed like the wrong operation here. I thought this was
> attempting to verify that the underlying WeakHandle has been released. But
> peeking doesn't ensure that. Oh, but we don't actually release() the
> WeakHandle when we "free" an OM. We're just pushing the OM on the free
> list. Which means the GC will continue to examine the associated OopStorage
> entry (and discover it's NULL). So there's some cost to not releasing the
> WeakHandle when putting an OM on the free list. Of course, it means there
> won't be any allocations when taking off the free list; I'm not sure which
> way is better.

Yeah, so we never really release the monitors, because the whole monitor can't be freed yet in this patch.

> But this makes me go back and wonder about whether object_peek() should have
> the _object.is_null() check. After creation it seems like that should never
> be true.

Before the switch away from TSM (next patch), we must assume that monitors have not yet been properly initialized, due to block allocation of a whole bunch of monitors at a time, before they have been associated with objects. They are also exposed to iterators and such, before any object association has been made. But very very soon, that will be a memory of the past.

> ------------------------------------------------------------------------------
> src/hotspot/share/runtime/synchronizer.cpp
> 
> The old code in chk_in_use_entry seems wrong. It checked for a null
> object() and recorded that as an error. But then it went on and attempted
> to use it as if it was not null. That's been fixed by the change. However,
> the change no longer treats a null as an error. Probably this is because
> it's weak, and so could have become null. But is that really possible for
> an "in use" monitor?

Well, the monitors are added to some in-use list at deflation time. The instant after, the monitor could have become unused. But it will still be floating around on the in-use list, until deflation time. And yes, now it is suddenly very much valid for a monitor that is no longer used (but placed on the in-use list when it was used), to have a NULL object, due to GC getting to weak processing clearing the oop, before deflation gets to it.

> ------------------------------------------------------------------------------
> src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ObjectMonitor.java
> 94 public OopHandle object() {
> 95 Address objAddr = addr.getAddressAt(objectFieldOffset);
> 96 if (objAddr == null) {
> 97 return null;
> 98 }
> 99 return objAddr.getOopHandleAt(0);
> 
> How about something a little bit less abstraction smashing?

This is already the "norm", and the way that all other handle loads look in the SA today, and therefore the SA does not support any GC with load barriers properly. While ideally, there would be something equivalent to the Access API in the SA to abstract the access details, I think it seems like way beyond the scope of this change to implement something like that for the SA, as part of this change.

Again, thanks for reviewing!

/Erik

fisk commented:
> Shenandoah changes are not complete:
> /home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp: In member function 'virtual void ShenandoahFinalMarkingTask::work(uint)':
> /home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:298:31: error: 'oops_do' is not a member of 'ObjectSynchronizer'
> ObjectSynchronizer::oops_do(&resolve_mark_cl);
> ^~~~~~~
> /home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:308:31: error: 'oops_do' is not a member of 'ObjectSynchronizer'
> ObjectSynchronizer::oops_do(&mark_cl);
> ^~~~~~~
> 
> I will have a look into how to resolve this.
> 
> In order to fix this, I need one of two things:
> 
>     * A way to iterate oops of the global-list. Explanation: In Shenandoah's I-U mode, we need to re-mark such ObjectMonitors during final-mark because they may be the only remaining reference to an object on the heap.
>       _or_ (even better)
> 
>     * Call a write-barrier (IN_NATIVE) whenever an ObjectMonitor is moved to the global-list upon thread-destruction. This way we could mark that object concurrently.
> 
> 
> Notice that this does not affect thread-local ObjectMonitors (IOW, most regular ObjectMonitors) because those would be scanned while scanning thread stacks. Therefore I'd like to avoid to generally place a barrier in ObjectMonitor.
> 
> See: https://bugs.openjdk.java.net/browse/JDK-8251451
> 
> AFAICT, this may affect ZGC too (not 100% sure about this).

So this change makes the object monitors weak. So you shouldn't have to remark them at all. If they hold the last reference to the object, then the monitor gets the object cleared as part of normal weak OopStorage reference processing. Subsequent deflation cycles will detect that the GC has cleared the oop, and reuse the monitor. In other words, we should just remove the call to the oops_do function, and rely on OopStorage doing its thing. The GC should not have to care at all about monitors any longer, only about processing weak OopStorages, which is done automatically. Hope this makes sense.

rkennke commented:
> @rkennke - Thanks for the review. I believe @fisk is going to address
> your comments.

Actually, if I understand it correctly, OopStorage already gives us full barriers, so we should be covered, and we can simply revert JDK-8251451:
http://cr.openjdk.java.net/~rkennke/8247281-shenandoah.patch
(Is there a better way to propose amendments to a PR?!)

I believe this probably also means that we don't need to scan object monitor lists during thread-scans, and let SATB (or whatever concurrent marking) take care of it. @fisk WDYT?

rkennke commented:
> > Shenandoah changes are not complete:
> > /home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp: In member function 'virtual void ShenandoahFinalMarkingTask::work(uint)':
> > /home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:298:31: error: 'oops_do' is not a member of 'ObjectSynchronizer'
> > ObjectSynchronizer::oops_do(&resolve_mark_cl);
> > ^~~~~~~
> > /home/rkennke/src/openjdk/jdk/src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp:308:31: error: 'oops_do' is not a member of 'ObjectSynchronizer'
> > ObjectSynchronizer::oops_do(&mark_cl);
> > ^~~~~~~
> > I will have a look into how to resolve this.
> > In order to fix this, I need one of two things:
> > ```
> > * A way to iterate oops of the global-list. Explanation: In Shenandoah's I-U mode, we need to re-mark such ObjectMonitors during final-mark because they may be the only remaining reference to an object on the heap.
> >   _or_ (even better)
> > 
> > * Call a write-barrier (IN_NATIVE) whenever an ObjectMonitor is moved to the global-list upon thread-destruction. This way we could mark that object concurrently.
> > ```
> > 
> > 
> > Notice that this does not affect thread-local ObjectMonitors (IOW, most regular ObjectMonitors) because those would be scanned while scanning thread stacks. Therefore I'd like to avoid to generally place a barrier in ObjectMonitor.
> > See: https://bugs.openjdk.java.net/browse/JDK-8251451
> > AFAICT, this may affect ZGC too (not 100% sure about this).
> 
> So this change makes the object monitors weak. So you shouldn't have to remark them at all. If they hold the last reference to the object, then the monitor gets the object cleared as part of normal weak OopStorage reference processing. Subsequent deflation cycles will detect that the GC has cleared the oop, and reuse the monitor. In other words, we should just remove the call to the oops_do function, and rely on OopStorage doing its thing. The GC should not have to care at all about monitors any longer, only about processing weak OopStorages, which is done automatically. Hope this makes sense.

Yes, definitely. I came to the same conclusion. Thank you!
With the patch amended:
http://cr.openjdk.java.net/~rkennke/8247281-shenandoah.patch

I'm good with the change.

fisk commented:
> Not sure about the JVM TI changes!

Here is a generic comment. You mention that the specification doesn't make it clear whether the roots reported are strong or weak. This is true. There is no mention about roots being strong or weak here. Notably, all listed roots, are strong though. I believe that with the chosen interpretation of "root", it is implied that it is strong, and hence why strong vs weak roots is not discussed. According to the memory management glossary (https://www.memorymanagement.org/glossary/r.html#glossary-r), this is the definition of a root:
"In tracing garbage collection, a root holds a reference or set of references to objects that are a priori reachable. The root set is used as the starting point in determining all reachable data."
Note that, this definition of root makes it implicit that it is also strong. And after a long discussion with GC nerds in Stockholm, this is in fact also why we have IN_NATIVE oop accesses, rather than IN_ROOT oop accesses. Because the opinions differed too much about what a "root" is. But to me it really does look like the JVMTI interpretation of a root, is a strong root, and hence that this is precisely what the text is talking about when mentioning roots, and hence also why all the enumerated roots, are indeed strong roots. Same comments for the HPROF_GC_ROOT_* discussion, as for the JVMTI_HEAP_ROOT_* discussion. That is also why I think it is now a bug to report them, as they are no longer "roots" in the JVMTI/HPROF sense. But you are right, filing a CSR for the behavioural change is probably a good idea.

fisk commented:
> > @rkennke - Thanks for the review. I believe @fisk is going to address
> > your comments.
> 
> Actually, if I understand it correctly, OopStorage already gives us full barriers, so we should be covered, and we can simply revert JDK-8251451:
> http://cr.openjdk.java.net/~rkennke/8247281-shenandoah.patch
> (Is there a better way to propose amendments to a PR?!)
> 
> I believe this probably also means that we don't need to scan object monitor lists during thread-scans, and let SATB (or whatever concurrent marking) take care of it. @fisk WDYT?

Absolutely. GC should no longer have to know anything about ObjectMonitors, only the automatically plugged in OopStorage that it processes. GC should not walk any monitor lists at all.

dcubed-ojdk commented:
@rkennke - I've commited the changes in your webrev as https://github.com/openjdk/jdk/pull/135/commits/bbf8dbd09bdf5c1c77c67cc637fbc10fe72d4894.

dcubed-ojdk commented:
@dholmes-ora and @fisk - I've taken a first pass at creating a CSR:
JDK-8253121 migrate ObjectMonitor::_object to OopStorage
https://bugs.openjdk.java.net/browse/JDK-8253121

Please look it over and feel free to edit as needed. Since I don't do
CSR's often, what I've done might be all wrong. :-)

dcubed-ojdk commented:
/label csr

dcubed-ojdk commented:
/csr

dcubed-ojdk commented:
@kimbarrett:

> src/hotspot/share/oops/weakHandle.cpp
> 36 WeakHandle::WeakHandle(OopStorage* storage, oop obj) :
> 37 _obj(storage->allocate()) {
> 38 assert(obj != NULL, "no need to create weak null oop");
> 
> Please format this differently so the ctor-init-list is more easily
> distinguished from the body. I don't care that much which of the several
> alternatives is used.

After discussion with Erik, I changed the indent on L37 from two space to four spaces.

dcubed-ojdk commented:
@kimbarrett 

> src/hotspot/share/runtime/objectMonitor.cpp
> 244 // Check that object() and set_object() are called from the right context:
> 245 static void check_object_context() {
> 
> This seems like checking we would normally only do in a debug build. Is this
> really supposed to be done in product builds too? (It's written to support
> that, just wondering if that's really what we want.) Maybe these aren't
> called very often so it doesn't matter? I also see that guarantee (rather
> than assert) is used a fair amount in this and related code.

I've changed check_object_context() to only be defined and called
when ASSERT is defined. I've also changed the guarantee() calls
to assert() calls.

I've done a couple of Mach5 Tier[1-8] test cycles on this code so I'm
no longer worried about this code or its callers in release bits.

dcubed-ojdk commented:
@kimbarrett 

> src/hotspot/share/runtime/objectMonitor.cpp
> 249 guarantee(self->is_Java_thread() || self->is_VM_thread(), "must be");
> 250 if (self->is_Java_thread()) {
> 
> Maybe instead
> 
> if (self->is_Java_thread()) {
> ...
> } else {
> guarantee(self->is_VM_thread(), "must be");
> }

I've made this refactoring change, tweaked the comments above
the block a bit and switched from guarantee() to assert().

dcubed-ojdk commented:
@kimbarrett 

> src/hotspot/share/runtime/synchronizer.cpp
> 1548 guarantee(m->object_peek() == NULL, "invariant");
> 
> Later: But see previous comment. Some of this might be relevat later though.
> 
> object_peek() seemed like the wrong operation here. I thought this was
> attempting to verify that the underlying WeakHandle has been released. But
> peeking doesn't ensure that. Oh, but we don't actually release() the
> WeakHandle when we "free" an OM. We're just pushing the OM on the free
> list. Which means the GC will continue to examine the associated OopStorage
> entry (and discover it's NULL). So there's some cost to not releasing the
> WeakHandle when putting an OM on the free list. Of course, it means there
> won't be any allocations when taking off the free list; I'm not sure which
> way is better.
> 
> But this makes me go back and wonder about whether object_peek() should have
> the _object.is_null() check. After creation it seems like that should never
> be true.

m->object_peek() == NULL is the right check at that location. om_release()
is called when we are returning an ObjectMonitor to a free list. At that point,
it should never be associated with an object.

dcubed-ojdk commented:
@kimbarrett 

> src/hotspot/share/runtime/synchronizer.cpp
> 
> The old code in chk_in_use_entry seems wrong. It checked for a null
> object() and recorded that as an error. But then it went on and attempted
> to use it as if it was not null. That's been fixed by the change. However,
> the change no longer treats a null as an error. Probably this is because
> it's weak, and so could have become null. But is that really possible for
> an "in use" monitor?

In the old code, an ObjectMonitor's object field should never be NULL when
that ObjectMonitor is on an in-use list. We'll get the logging message and
then a crash. I used to have guarantee(n->object() != NULL, ...) in there,
but Robbin convinced me that was a waste because we'll just crash on
the use of the NULL pointer and that was good enough.

As Erik has already explained, now that we use a weak handle, the object
can be GC'ed before the deflater thread comes along and removes the
idle ObjectMonitor from the in-use list.

dcubed-ojdk commented:
@kimbarrett - I believe I've addressed your comments in this push:
https://github.com/openjdk/jdk/pull/135/commits/9fa2bed109d6fe352c610b26ada650226ce9cec4

dcubed-ojdk commented:
@coleenp, @rkennke, and @kimbarrett - I believe all of the changes that you have requested
have been made. Please confirm by re-reviewing.

@dholmes-ora - I don't think you asked for any specific code changes. I've taken a first pass
at creating a CSR:
JDK-8253121 migrate ObjectMonitor::_object to OopStorage
https://bugs.openjdk.java.net/browse/JDK-8253121

Please look it over and feel free to edit as needed. Since I don't do
CSR's often, what I've done might be all wrong. :-)

rkennke commented:
> @coleenp, @rkennke, and @kimbarrett - I believe all of the changes that you have requested
> have been made. Please confirm by re-reviewing.

It looks good to me now! I've also build with Shenandoah GC, and run our sanity-tests (TEST=hotspot_gc_shenandoah) and all looks good. Thank you!

rkennke commented on src/hotspot/share/gc/shared/space.inline.hpp:
```diff
@@ -175,6 +171,9 @@ inline void CompactibleSpace::scan_and_forward(SpaceType* space, CompactPoint* c
     } else {
+      assert(!space->scanned_block_is_obj(cur_obj) || oop(cur_obj)->mark_raw().is_unlocked() ||
+             oop(cur_obj)->mark_raw().has_bias_pattern() || oop(cur_obj)->mark_raw().has_monitor(),
+             "these are the only valid states during a mark sweep");
```
Is this change related? Also, when moving the assert into the else block it will become always-true because of space->scanned_block_is_obj(cur_obj), or am I missing something?

dcubed-ojdk replied:
See this comment from Coleen and the replies:
https://github.com/openjdk/jdk/pull/135#discussion_r487300636

Please let me know if that resolved this comment for you.

rkennke replied:
Sorry, no. Maybe it's too late here and I shall think about it tomorrow morning instead ;-) Or maybe you can explain it again in the context of that change. How's the assert even relevant when moved in the else-branch?

dcubed-ojdk replied:
Sorry, I confused myself switching between this review and a
preliminary review thread.

Here's the original code:

165   while (cur_obj < scan_limit) {
166     assert(!space->scanned_block_is_obj(cur_obj) ||
167            oop(cur_obj)->mark_raw().is_marked() || oop(cur_obj)->mark_raw().is_unlocked() ||
168            oop(cur_obj)->mark_raw().has_bias_pattern(),
169            "these are the only valid states during a mark sweep");
170     if (space->scanned_block_is_obj(cur_obj) && oop(cur_obj)->is_gc_marked()) {

and here's the code after it was moved and rewritten:

173     } else {
174       assert(!space->scanned_block_is_obj(cur_obj) || oop(cur_obj)->mark_raw().is_unlocked() ||
175              oop(cur_obj)->mark_raw().has_bias_pattern() || oop(cur_obj)->mark_raw().has_monitor(),
176              "these are the only valid states during a mark sweep");


Where the assert() was worked fine when the ObjectMonitor had a regular oop,
but after it was changed into a weak handle, that location became exposed to
the fact that the object reference could be GC'ed. The original code assumes
that the ObjectMonitor oop reference is stable and unchanging and that's no
longer the case with the weak handle.

dcubed-ojdk replied:
I found my original preliminary code review comment and Erik's reply:

> src/hotspot/share/gc/shared/space.inline.hpp
>     old L166:     assert(!space->scanned_block_is_obj(cur_obj) ||
>     old L167:            oop(cur_obj)->mark_raw().is_marked() || oop(cur_obj)->mark_raw().is_unlocked() ||
>     old L168: oop(cur_obj)->mark_raw().has_bias_pattern(),
>     old L169:            "these are the only valid states during a mark sweep");
>         This assert was before the if-statement at the top of the while-loop.
>
>     new L174: assert(!space->scanned_block_is_obj(cur_obj) || oop(cur_obj)->mark_raw().is_unlocked() ||
>     new L175: oop(cur_obj)->mark_raw().has_bias_pattern() || oop(cur_obj)->mark_raw().has_monitor(),
>     new L176:              "these are the only valid states during a mark sweep");
>         The assert is now in the else branch of the following if-statement:
>
>            L166     if (space->scanned_block_is_obj(cur_obj) && oop(cur_obj)->is_gc_marked()) {
>
>         The new assert() drops this check:
>
>             oop(cur_obj)->mark_raw().is_marked()
>
>         and adds this check:
>
>             oop(cur_obj)->mark_raw().has_monitor()
>
>         Dropping the "is_marked()" makes sense since the new location
>         of the assert is in the else branch of "oop(cur_obj)->is_gc_marked()".
>         The addition of the "has_monitor()" check is puzzling. Why was this
>         added and why wasn't it needed in the old assert()?
>
>         In fact, I'm not sure why this change is here at all.

This is an artifact of the monitor now being weak. Since there was previously always a strong root
to all inflated monitors, there were never any dead objects in the heap, that still had pointers in the
mark word to the monitor. The change to weak now implies that we suddenly have dead objects
in the heap, that in the markWord point out their monitor. GC code that iterates through consecutive
objects one by one, will see these now dead objects with monitors. The assert changes reflect that.
Before it was unexpected and would assert on that. Now I moved the assertion to the case when the
object is alive instead. We have no business asserting what should be in the markWord of dead objects.

I hope it makes sense now!

coleenp replied:
This code seems like something that doesn't belong here anymore.  This code assumed synchronous scanning of oops in ObjectMonitor and scanning memory regions, and that's no longer the case with OopStorage. I think this assert should be removed.  It exports some implementation detail of now completely unrelated code in order to do a very specific check.

dcubed-ojdk replied:
@fisk - please chime in here...

rkennke replied:
I agree. Also, the assert becomes true somewhat obviously because of its first clause, which is already guaranteed because of its placement in the surrounding else-branch (unless something really weird happens).

fisk replied:
I am also 100% okay with just removing the random assertion. We have a whole bunch of other block iteration code in GC code that does not sprinkle random asserts about what patterns of markWords are supposedly good or bad in live/dead objects. And I also don't get what kind of bug this would catch.

dcubed-ojdk replied:
Okay. Since @coleenp, @rkennke and @fisk want that assertion to "hit the road",
I'm in the process of removing it... build running now (just 'cause I'm paranoid me...)...

dcubed-ojdk replied:
This sub-thread should be resolve via https://github.com/openjdk/jdk/pull/135/commits/eeb9d761374ea193a7295fcce3ffe9707a8f0348

dcubed-ojdk commented:
@rkennke - Thanks for confirming that Shenandoah now builds with the changes
that you provided. Thanks for sending those.

dholmes-ora commented:
/csr needed

kimbarrett commented on src/hotspot/share/runtime/objectMonitor.cpp:
```diff
@@ -241,13 +241,12 @@ void ObjectMonitor::operator delete[] (void *p) {
   operator delete(p);
 }
 
+#ifdef ASSERT
```
There would be less `#ifdef ASSERT` clutter if just the body of check_object_context were conditionalized.  Then the calls wouldn't need to be.  Your call...

dcubed-ojdk replied:
I'll make that change.

coleenp reviewed (approved):
Thank you for removing it.

rkennke reviewed (approved):
Looks good to me!

dcubed-ojdk commented:
@kimbarrett - Thanks for the cleanup suggestion.
@sspitsyn - Thanks for the review.
I've gone ahead and "resolved" the comments that were addressed by previous commits.
I've made changes for both Kim and Serguei's reviews. Building those bits now.

dcubed-ojdk commented:
@kimbarrett and @sspitsyn - Your most recent comments should be resolved via https://github.com/openjdk/jdk/pull/135/commits/215084ac7ef481713560d14498ce420a40ca813a.

dcubed-ojdk commented:
/integrate
